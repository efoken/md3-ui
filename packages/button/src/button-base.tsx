import { useEventCallback, useForkRef } from "@md3-ui/hooks"
import {
  OwnerStateProps,
  styled,
  StyleSheet,
  SxProps,
  useTheme,
  useThemeProps,
} from "@md3-ui/system"
import { splitProps, __DEV__ } from "@md3-ui/utils"
import * as React from "react"
import {
  Platform,
  Pressable as RNPressable,
  PressableProps as RNPressableProps,
  StyleProp,
  View as RNView,
  ViewStyle as RNViewStyle,
} from "react-native"

export type ButtonBaseProps = Omit<
  RNPressableProps,
  "children" | "disabled" | "style"
> & {
  /** @default false */
  centerRipple?: boolean
  children?: React.ReactNode
  /** @default false */
  disabled?: boolean
  /** @default false */
  disableRipple?: boolean
  focusColor?: string
  /** @default 0.12 */
  focusOpacity?: string
  hoverColor?: string
  /** @default 0.08 */
  hoverOpacity?: number
  pressedColor?: string
  /** @default 0.12 */
  pressedOpacity?: number
  rippleColor?: string
  style?: StyleProp<RNViewStyle>
  styles?: {
    root?: RNViewStyle
    container?: RNViewStyle
  }
  sx?: SxProps
}

export type ButtonBaseStyleKey = keyof NonNullable<ButtonBaseProps["styles"]>

const ButtonBaseRoot = styled(RNPressable, {
  name: "ButtonBase",
  slot: "Root",
})<OwnerStateProps<Pick<ButtonBaseProps, "disabled">>>(
  ({ ownerState }) =>
    Platform.OS === "web" && {
      cursor: ownerState.disabled ? "default" : "pointer",
      outlineWidth: 0,
      overflow: "hidden",
      transition: "box-shadow 200ms linear",
      userSelect: "none",
    },
)

const ButtonBaseContainer = styled(RNView, {
  name: "ButtonBase",
  slot: "Container",
})({
  backgroundColor: "rgba(255, 255, 255, 0)",
  overflow: "hidden",
})

const ButtonBaseHover = styled("span", {
  name: "ButtonBase",
  slot: "Hover",
})({
  bottom: 0,
  left: 0,
  position: "absolute",
  right: 0,
  top: 0,
  transition: "opacity 200ms linear",
  zIndex: -1,
})

const ButtonBaseRippleContainer = styled("span", {
  name: "ButtonBase",
  slot: "RippleContainer",
})()

const ButtonBaseRipple = styled("span", {
  name: "ButtonBase",
  slot: "Ripple",
})<
  OwnerStateProps<
    Required<
      Pick<ButtonBaseProps, "hoverOpacity" | "pressedColor" | "pressedOpacity">
    >
  >
>(({ ownerState, theme }) => ({
  animationDuration: "550ms",
  animationKeyframes: {
    "0%": {
      opacity: 1,
      transform: [{ scale: 0 }],
    },
    "100%": {
      opacity: 0,
      transform: [{ scale: 1 }],
    },
  },
  animationTimingFunction: "ease-in",
  backgroundColor: theme.utils.rgba(
    ownerState.pressedColor,
    ownerState.pressedOpacity - ownerState.hoverOpacity,
  ),
  borderRadius: 999,
  pointerEvents: "none",
  position: "absolute",
  zIndex: -1,
}))

export const ButtonBase = React.forwardRef<RNView, ButtonBaseProps>(
  (inProps, ref) => {
    const {
      centerRipple = false,
      children,
      disabled = false,
      disableRipple = false,
      focusColor: focusColorProp,
      focusOpacity = 0.12,
      hoverColor: hoverColorProp,
      hoverOpacity = 0.08,
      href,
      onKeyDown,
      onKeyUp,
      onPress,
      pressedColor: pressedColorProp,
      pressedOpacity = 0.12,
      rippleColor,
      style,
      styles,
      ...props
    } = useThemeProps({ name: "ButtonBase", props: inProps })

    const theme = useTheme()

    const rootRef = React.useRef<RNView>(null)
    const handleRef = useForkRef(rootRef, ref)

    const [ripples, setRipples] = React.useState<
      { id: number; ripple: React.ReactElement }[]
    >([])

    const [{ backgroundColor = null, ...containerStyle } = {}] = splitProps(
      StyleSheet.flatten(style),
      [
        "backgroundColor",
        "borderRadius",
        "bottom",
        "height",
        "left",
        "margin",
        "marginBottom",
        "marginEnd",
        "marginHorizontal",
        "marginLeft",
        "marginRight",
        "marginStart",
        "marginTop",
        "marginVertical",
        "position",
        "right",
        "top",
        "width",
      ],
    )

    const hoverColor = rippleColor ?? hoverColorProp ?? theme.color.background
    const focusColor = rippleColor ?? focusColorProp ?? hoverColor
    const pressedColor =
      rippleColor ?? pressedColorProp ?? theme.color.background

    const underlayColor =
      // When button has no background color, simply use `rippleColor` as is
      backgroundColor == null
        ? theme.utils.rgba(pressedColor, pressedOpacity)
        : // Otherwise we mix the `rippleColor` with the background color
          theme.utils.mix(
            pressedOpacity,
            pressedColor,
            backgroundColor.toString(),
          )

    const removeRipple = (rippleID: number) => () => {
      setRipples((prevRipples) => prevRipples.filter((r) => r.id !== rippleID))
    }

    const appendRipple = React.useCallback(
      (event: TouchEvent | MouseEvent | {} = {}) => {
        const currentRef = rootRef.current as HTMLDivElement | null

        if (disabled || disableRipple || currentRef == null) {
          return
        }

        const { width, height, left, top } = currentRef.getBoundingClientRect()

        // Get the size of the ripple
        let rippleX: number
        let rippleY: number
        let rippleSize: number

        if (
          centerRipple ||
          (!("clientX" in event) && !("touches" in event)) ||
          ("clientX" in event && event?.clientX === 0 && event?.clientY === 0)
        ) {
          rippleX = Math.round(width / 2)
          rippleY = Math.round(height / 2)
        } else {
          const { clientX, clientY } =
            "touches" in event ? event.touches[0] : event
          rippleX = Math.round(clientX - left)
          rippleY = Math.round(clientY - top)
        }

        if (centerRipple) {
          rippleSize = Math.sqrt((2 * width ** 2 + height ** 2) / 3)
        } else {
          const sizeX =
            Math.max(Math.abs(currentRef.clientWidth - rippleX), rippleX) * 2 +
            2
          const sizeY =
            Math.max(Math.abs(currentRef.clientHeight - rippleY), rippleY) * 2 +
            2
          rippleSize = Math.sqrt(sizeX ** 2 + sizeY ** 2)
        }

        const rippleID =
          ripples.length === 0 ? 1 : ripples[ripples.length - 1].id + 1

        const ripple = (
          <ButtonBaseRipple
            key={rippleID}
            ownerState={{
              hoverOpacity,
              pressedColor,
              pressedOpacity,
            }}
            style={{
              height: rippleSize,
              left: -(rippleSize / 2) + rippleX,
              top: -(rippleSize / 2) + rippleY,
              width: rippleSize,
            }}
            onAnimationEnd={removeRipple(rippleID)}
          />
        )

        setRipples((prevRipples) => [...prevRipples, { id: rippleID, ripple }])
      },
      [
        rootRef,
        disabled,
        disableRipple,
        centerRipple,
        ripples,
        pressedColor,
        pressedOpacity,
        hoverOpacity,
      ],
    )

    const isNonNativeButton = () => {
      const buttonEl = rootRef.current as HTMLElement | null
      return (
        buttonEl?.tagName !== "BUTTON" &&
        !(buttonEl?.tagName === "A" && (buttonEl as HTMLAnchorElement).href)
      )
    }

    const handleKeyDown = useEventCallback((event: React.KeyboardEvent) => {
      if (event.key === "Enter" || event.key === " ") {
        appendRipple()
      }

      if (
        event.target === event.currentTarget &&
        isNonNativeButton() &&
        event.key === " "
      ) {
        event.preventDefault()
      }

      onKeyDown?.(event)

      // Keyboard accessibility for non interactive elements
      if (
        event.target === event.currentTarget &&
        isNonNativeButton() &&
        event.key === "Enter" &&
        !disabled
      ) {
        event.preventDefault()
        onPress?.(event as any)
      }
    })

    const handleKeyUp = useEventCallback((event: React.KeyboardEvent) => {
      onKeyUp?.(event)

      // Keyboard accessibility for non interactive elements
      if (
        event.target === event.currentTarget &&
        isNonNativeButton() &&
        event.key === " " &&
        !event.defaultPrevented
      ) {
        onPress?.(event as any)
      }
    })

    React.useEffect(() => {
      const currentRef = rootRef.current as HTMLDivElement | null

      if (Platform.OS === "web" && currentRef != null) {
        currentRef.addEventListener("mousedown", appendRipple)
        currentRef.addEventListener("touchstart", appendRipple)

        return () => {
          currentRef.removeEventListener("mousedown", appendRipple)
          currentRef.removeEventListener("touchend", appendRipple)
        }
      }

      return () => {}
    }, [appendRipple, handleKeyDown, rootRef])

    const ownerState = {
      disabled,
    }

    const button = (
      <ButtonBaseRoot
        ref={handleRef}
        accessibilityRole={href ? undefined : "button"}
        accessibilityState={{ disabled: disabled || undefined }}
        android_ripple={
          disableRipple
            ? {}
            : {
                borderless: true,
                color: pressedColor,
              }
        }
        disabled={disabled}
        href={href}
        ownerState={ownerState}
        pointerEvents={disabled ? "none" : undefined}
        {...(Platform.OS === "web" && {
          style: [style, styles?.root],
        })}
        {...(Platform.OS === "android" && {
          // For Android we need a wrapping View to cut off the ripple effect.
          // Because of this View and when the button should have
          // `position: absolute`, we need to apply it to the wrapper View and
          // override it here to have `position: relative` with 0px inset.
          style: [
            style,
            styles?.root,
            {
              bottom: 0,
              left: 0,
              marginBottom: 0,
              marginEnd: 0,
              marginLeft: 0,
              marginRight: 0,
              marginStart: 0,
              marginTop: 0,
              position: "relative",
              right: 0,
              top: 0,
            },
          ],
        })}
        {...(Platform.OS === "ios" && {
          style: ({ pressed }) => [
            style,
            styles?.root,
            pressed && !disableRipple && { backgroundColor: underlayColor },
          ],
        })}
        onKeyDown={handleKeyDown}
        onKeyUp={handleKeyUp}
        onPress={onPress}
        {...props}
      >
        {({ hovered, focused }) => (
          <>
            {React.Children.only(children)}
            {Platform.OS === "web" && (
              <ButtonBaseHover
                style={{
                  backgroundColor: focused ? focusColor : hoverColor,
                  opacity: focused ? focusOpacity : hovered ? hoverOpacity : 0,
                }}
              />
            )}
            {Platform.OS === "web" && !disableRipple && (
              <ButtonBaseRippleContainer>
                {ripples.map(({ ripple }) => ripple)}
              </ButtonBaseRippleContainer>
            )}
          </>
        )}
      </ButtonBaseRoot>
    )

    return Platform.OS === "android" ? (
      <ButtonBaseContainer style={[containerStyle, styles?.container]}>
        {button}
      </ButtonBaseContainer>
    ) : (
      button
    )
  },
)

if (__DEV__) {
  ButtonBase.displayName = "ButtonBase"
}
