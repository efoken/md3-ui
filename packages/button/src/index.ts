export { ButtonBase } from "./button-base"
export type {
  ButtonBaseProps,
  ButtonBaseStateCallbackType,
  ButtonBaseStyleKey,
} from "./button-base"
export { ElevatedButton } from "./elevated-button"
export type {
  ElevatedButtonProps,
  ElevatedButtonStyleKey,
} from "./elevated-button"
export { ExtendedFab } from "./extended-fab"
export type { ExtendedFabProps, ExtendedFabStyleKey } from "./extended-fab"
export { Fab } from "./fab"
export type { FabProps, FabStyleKey } from "./fab"
export { FilledButton } from "./filled-button"
export type { FilledButtonProps, FilledButtonStyleKey } from "./filled-button"
export * from "./filled-icon-button"
export * from "./icon-button"
export * from "./outlined-button"
export * from "./outlined-icon-button"
export * from "./text-button"
export * from "./tonal-button"
export * from "./tonal-icon-button"
export * from "./use-button-base-state"
