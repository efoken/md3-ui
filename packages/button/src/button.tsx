import { useBoolean } from "@md3-ui/hooks"
import { Text, TextProvider } from "@md3-ui/layout"
import {
  OwnerStateProps,
  styled,
  SxProps,
  useTheme,
  useThemeProps,
} from "@md3-ui/system"
import { __DEV__ } from "@md3-ui/utils"
import * as React from "react"
import {
  TextStyle as RNTextStyle,
  View as RNView,
  ViewStyle as RNViewStyle,
} from "react-native"
import { ButtonBase, ButtonBaseProps } from "./button-base"

export interface ButtonProps extends ButtonBaseProps {
  /**
   * The content of the component.
   */
  children?: React.ReactNode
  /**
   * If `true`, the component is disabled.
   * @default false
   */
  disabled?: boolean
  /**
   * Element placed before the children.
   */
  icon?: React.ReactElement
  /**
   * Override or extend the styles applied to the component.
   */
  styles?: {
    root?: RNViewStyle
    content?: RNTextStyle
    icon?: RNViewStyle
    label?: RNTextStyle
  }
  /**
   * The system prop that allows defining system overrides as well as
   * additional styles.
   */
  sx?: SxProps
  /**
   * The variant to use.
   * @default "elevated"
   */
  variant?: "elevated" | "filled" | "tonal" | "outlined" | "text"
}

export type ButtonStyleKey = keyof NonNullable<ButtonProps["styles"]>

const ButtonRoot = styled(ButtonBase, {
  name: "Button",
  slot: "Root",
})<OwnerStateProps<Pick<ButtonProps, "disabled" | "icon" | "variant">>>(
  ({ theme, ownerState }) => [
    {
      alignItems: "center",
      borderRadius: 20,
      flexDirection: "row",
      height: 40,
      paddingHorizontal: theme.spacing(3),

      ...(ownerState.variant === "elevated" && {
        ...theme.elevation.level1,
        backgroundColor: theme.color.surface,

        ...(ownerState.disabled && {
          ...theme.elevation.level0,
          backgroundColor: theme.utils.rgba(theme.color["on-surface"], 0.12),
        }),

        ":hover": {
          ...theme.elevation.level2,
        },

        ":focus": {
          ...theme.elevation.level1,
        },
      }),

      ...(ownerState.variant === "filled" && {
        ...theme.elevation.level0,
        backgroundColor: theme.color.primary,

        ...(ownerState.disabled && {
          ...theme.elevation.level0,
          backgroundColor: theme.utils.rgba(theme.color["on-surface"], 0.12),
        }),

        ":hover": {
          ...theme.elevation.level1,
        },

        ":focus": {
          ...theme.elevation.level0,
        },
      }),

      ...(ownerState.variant === "tonal" && {
        ...theme.elevation.level0,
        backgroundColor: theme.color["secondary-container"],

        ...(ownerState.disabled && {
          backgroundColor: theme.utils.rgba(theme.color["on-surface"], 0.12),
        }),

        ":hover": {
          ...theme.elevation.level1,
        },

        ":focus": {
          ...theme.elevation.level0,
        },
      }),

      ...(ownerState.variant === "outlined" && {
        ...theme.elevation.level0,
        backgroundColor: "rgba(255, 255, 255, 0)",
        borderColor: theme.color.outline,
        borderWidth: 1,
        paddingHorizontal: theme.spacing(3) - 1,

        ...(ownerState.disabled && {
          borderColor: theme.utils.rgba(theme.color["on-surface"], 0.12),
        }),

        ":focus": {
          borderColor: theme.color.primary,
        },
      }),

      ...(ownerState.variant === "text" && {
        ...theme.elevation.level0,
        minWidth: 48,
        paddingEnd: ownerState.icon ? theme.spacing(2) : theme.spacing(1.5),
        paddingStart: theme.spacing(1.5),
      }),
    },
  ],
)

const ButtonContent = styled(TextProvider, {
  name: "Button",
  slot: "Content",
})<OwnerStateProps<Pick<ButtonProps, "disabled" | "variant">>>(
  ({ theme, ownerState }) => ({
    ...(ownerState.variant === "elevated" && {
      color: theme.color.primary,
    }),

    ...(ownerState.variant === "filled" && {
      color: theme.color["on-primary"],
    }),

    ...(ownerState.variant === "tonal" && {
      color: theme.color["on-secondary-container"],
    }),

    ...(ownerState.variant === "outlined" && {
      color: theme.color.primary,
    }),

    ...(ownerState.variant === "text" && {
      color: theme.color.primary,
    }),

    ...(ownerState.disabled && {
      color: theme.utils.rgba(theme.color["on-surface"], 0.38),
    }),
  }),
)

const ButtonIcon = styled(RNView, {
  name: "Button",
  slot: "Icon",
})<OwnerStateProps<Pick<ButtonProps, "variant">>>(({ theme, ownerState }) => ({
  marginEnd: theme.spacing(1),
  marginStart: ownerState.variant === "text" ? 0 : -theme.spacing(1),
}))

const ButtonLabel = styled(Text, {
  name: "Button",
  slot: "Label",
})(({ theme }) => ({
  ...theme.typescale["label-large"],
  textAlign: "center",
}))

export const Button = React.forwardRef<RNView, ButtonProps>((inProps, ref) => {
  const {
    children,
    disabled = false,
    icon,
    style,
    styles,
    variant = "elevated",
    ...props
  } = useThemeProps({ name: "Button", props: inProps })

  const theme = useTheme()

  const [hovered, handleHover] = useBoolean()
  const [focused, handleFocus] = useBoolean()

  const ownerState = {
    disabled,
    focused,
    hovered,
    icon,
    variant,
  }

  return (
    <ButtonRoot
      ref={ref}
      disabled={disabled}
      ownerState={ownerState}
      rippleColor={
        variant === "elevated"
          ? theme.color.primary
          : variant === "filled"
          ? theme.color["on-primary"]
          : variant === "tonal"
          ? theme.color["on-secondary-container"]
          : variant === "outlined"
          ? theme.color.primary
          : variant === "text"
          ? theme.color.primary
          : undefined
      }
      style={[styles?.root, style]}
      onBlur={handleFocus.off}
      onFocus={handleFocus.on}
      onHoverIn={handleHover.on}
      onHoverOut={handleHover.off}
      {...props}
    >
      <ButtonContent ownerState={ownerState} style={styles?.content}>
        {icon && (
          <ButtonIcon ownerState={ownerState} style={styles?.icon}>
            {React.cloneElement(icon, {
              height: 18,
              width: 18,
            })}
          </ButtonIcon>
        )}
        <ButtonLabel style={styles?.label}>{children}</ButtonLabel>
      </ButtonContent>
    </ButtonRoot>
  )
})

if (__DEV__) {
  Button.displayName = "Button"
}
