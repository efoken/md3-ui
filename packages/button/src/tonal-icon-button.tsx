import {
  OwnerStateProps,
  styled,
  StylesProp,
  SxProps,
  TextStyleProvider,
  useTheme,
  useThemeProps,
} from "@md3-ui/system"
import * as React from "react"
import {
  TextStyle as RNTextStyle,
  View as RNView,
  ViewStyle as RNViewStyle,
} from "react-native"
import { ButtonBase, ButtonBaseProps } from "./button-base"
import { useButtonBaseState } from "./use-button-base-state"

export interface TonalIconButtonProps extends ButtonBaseProps {
  /**
   * The icon to display.
   */
  children: React.ReactElement
  /**
   * If given, uses a negative margin to counteract the padding on one side
   * (this is often helpful for aligning the left or right side of the icon with
   * content above or below, without ruining the border size and shape).
   */
  edge?: "start" | "end" | false
  /**
   * If `true`, the component is selected.
   */
  selected?: boolean
  /**
   * Override or extend the styles applied to the component.
   */
  styles?: StylesProp<{
    root?: RNViewStyle
    content?: RNTextStyle
  }>
  /**
   * The system prop that allows defining system overrides as well as
   * additional styles.
   */
  sx?: SxProps
  /** @default false */
  toggle?: boolean
}

export type TonalIconButtonStyleKey = keyof NonNullable<
  TonalIconButtonProps["styles"]
>

type TonalIconButtonOwnerState = Pick<
  TonalIconButtonProps,
  "disabled" | "edge" | "selected" | "toggle"
> & {
  focused: boolean
  hovered: boolean
  pressed: boolean
}

const TonalIconButtonRoot = styled(ButtonBase, {
  name: "TonalIconButton",
  slot: "Root",
})<OwnerStateProps<TonalIconButtonOwnerState>>(({ theme, ownerState }) => ({
  alignItems: "center",
  backgroundColor: theme.comp.tonalIconButton.container.color,
  borderRadius: theme.comp.tonalIconButton.container.shape,
  height: theme.comp.tonalIconButton.container.size,
  justifyContent: "center",
  marginEnd: ownerState.edge === "end" ? -12 : undefined,
  marginStart: ownerState.edge === "start" ? -12 : undefined,
  width: theme.comp.tonalIconButton.container.size,

  ...(ownerState.toggle && {
    backgroundColor: ownerState.selected
      ? theme.comp.tonalIconButton.selected.container.color
      : theme.comp.tonalIconButton.unselected.container.color,
  }),

  ...(ownerState.disabled && {
    backgroundColor: theme.utils.rgba(
      theme.comp.tonalIconButton.disabled.container.color,
      theme.comp.tonalIconButton.disabled.container.opacity,
    ),
  }),
}))

const TonalIconButtonContent = styled(TextStyleProvider, {
  name: "TonalIconButton",
  slot: "Content",
  skipSx: true,
})<OwnerStateProps<TonalIconButtonOwnerState>>(({ theme, ownerState }) => ({
  color: theme.comp.tonalIconButton.icon.color,

  ...(ownerState.hovered && {
    color: theme.comp.tonalIconButton.hover.icon.color,
  }),

  ...(ownerState.focused && {
    color: theme.comp.tonalIconButton.focus.icon.color,
  }),

  ...(ownerState.pressed && {
    color: theme.comp.tonalIconButton.pressed.icon.color,
  }),

  ...(ownerState.toggle && {
    color: ownerState.selected
      ? theme.comp.tonalIconButton.toggle.selected.icon.color
      : theme.comp.tonalIconButton.toggle.unselected.icon.color,

    ...(ownerState.hovered && {
      color: ownerState.selected
        ? theme.comp.tonalIconButton.toggle.selected.hover.icon.color
        : theme.comp.tonalIconButton.toggle.unselected.hover.icon.color,
    }),

    ...(ownerState.focused && {
      color: ownerState.selected
        ? theme.comp.tonalIconButton.toggle.selected.focus.icon.color
        : theme.comp.tonalIconButton.toggle.unselected.focus.icon.color,
    }),

    ...(ownerState.pressed && {
      color: ownerState.selected
        ? theme.comp.tonalIconButton.toggle.selected.pressed.icon.color
        : theme.comp.tonalIconButton.toggle.unselected.pressed.icon.color,
    }),
  }),

  ...(ownerState.disabled && {
    color: theme.utils.rgba(
      theme.comp.tonalIconButton.disabled.icon.color,
      theme.comp.tonalIconButton.disabled.icon.opacity,
    ),
  }),
}))

export const TonalIconButton = React.forwardRef<RNView, TonalIconButtonProps>(
  (inProps, ref) => {
    const {
      children,
      disabled = false,
      edge = false,
      selected,
      style,
      styles,
      toggle = false,
      ...props
    } = useThemeProps({
      name: "TonalIconButton",
      props: inProps,
    })

    const theme = useTheme()

    const { focused, hovered, pressed, ...buttonBaseProps } =
      useButtonBaseState(props)

    const ownerState = {
      disabled,
      edge,
      focused,
      hovered,
      pressed,
      selected,
      toggle,
    }

    return (
      <TonalIconButtonRoot
        ref={ref}
        disabled={disabled}
        focusColor={theme.comp.tonalIconButton.focus.stateLayer.color}
        focusOpacity={theme.comp.tonalIconButton.focus.stateLayer.opacity}
        hitSlop={4}
        hoverColor={theme.comp.tonalIconButton.hover.stateLayer.color}
        hoverOpacity={theme.comp.tonalIconButton.hover.stateLayer.opacity}
        ownerState={ownerState}
        pressedColor={theme.comp.tonalIconButton.pressed.stateLayer.color}
        pressedOpacity={theme.comp.tonalIconButton.pressed.stateLayer.opacity}
        {...(toggle && {
          focusColor: selected
            ? theme.comp.tonalIconButton.toggle.selected.focus.stateLayer.color
            : theme.comp.tonalIconButton.toggle.unselected.focus.stateLayer
                .color,
          hoverColor: selected
            ? theme.comp.tonalIconButton.toggle.selected.hover.stateLayer.color
            : theme.comp.tonalIconButton.toggle.unselected.hover.stateLayer
                .color,
          pressedColor:
            theme.comp.tonalIconButton.toggle.unselected.pressed.stateLayer
              .color,
        })}
        style={[style, styles?.root]}
        {...buttonBaseProps}
        {...props}
      >
        <TonalIconButtonContent ownerState={ownerState} style={styles?.content}>
          {React.cloneElement(children, {
            height: theme.comp.tonalIconButton.icon.size,
            width: theme.comp.tonalIconButton.icon.size,
          })}
        </TonalIconButtonContent>
      </TonalIconButtonRoot>
    )
  },
)

TonalIconButton.displayName = "TonalIconButton"
