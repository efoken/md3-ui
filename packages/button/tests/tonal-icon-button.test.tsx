import { Icon } from "@md3-ui/icon"
import { describeConformance, render, screen } from "@md3-ui/test-utils"
import { TonalIconButton } from "../src/tonal-icon-button"

describe("<TonalIconButton />", () => {
  describeConformance(
    <TonalIconButton>
      <Icon testID="icon" />
    </TonalIconButton>,
    () => ({
      inheritComponent: "button",
      refInstanceof: window.HTMLButtonElement,
      skip: ["asProp"],
    }),
  )

  it("should render", () => {
    render(
      <TonalIconButton>
        <Icon testID="icon" />
      </TonalIconButton>,
    )
    const button = screen.getByRole("button")
    const icon = screen.getByTestId("icon")

    expect(button).toMatchSnapshot()
    expect(icon).not.toBeNull()
  })

  it("should have a ripple by default", () => {
    render(
      <TonalIconButton>
        <Icon />
      </TonalIconButton>,
    )
    const button = screen.getByRole("button")

    expect(button.querySelector(".ButtonBase-rippleContainer")).not.toBeNull()
  })

  it("should disable the ripple", () => {
    render(
      <TonalIconButton disableRipple>
        <Icon />
      </TonalIconButton>,
    )
    const button = screen.getByRole("button")

    expect(button.querySelector(".ButtonBase-rippleContainer")).toBeNull()
  })

  describe("prop: edge", () => {
    it('should render the correct margin on edge="start"', () => {
      const { container } = render(
        <TonalIconButton edge="start">
          <Icon />
        </TonalIconButton>,
      )

      expect(container.firstChild).toHaveStyle({
        marginLeft: -12,
      })
    })

    it('should render the correct margin on edge="end"', () => {
      const { container } = render(
        <TonalIconButton edge="end">
          <Icon />
        </TonalIconButton>,
      )

      expect(container.firstChild).toHaveStyle({
        marginRight: -12,
      })
    })

    it("should render no margin without edge", () => {
      const { container } = render(
        <TonalIconButton>
          <Icon />
        </TonalIconButton>,
      )

      expect(container.firstChild).toHaveStyle({
        marginLeft: 0,
        marginRight: 0,
      })
    })
  })

  describe("prop: selected", () => {
    it("should select the component", () => {
      render(
        <TonalIconButton toggle selected>
          <Icon />
        </TonalIconButton>,
      )
      const button = screen.getByRole("button")

      expect(button).toMatchSnapshot()
    })

    it("should unselect the component", () => {
      render(
        <TonalIconButton toggle selected={false}>
          <Icon />
        </TonalIconButton>,
      )
      const button = screen.getByRole("button")

      expect(button).toMatchSnapshot()
    })
  })

  describe("prop: disabled", () => {
    it("should disable the component", () => {
      render(
        <TonalIconButton disabled>
          <Icon />
        </TonalIconButton>,
      )
      const button = screen.getByRole("button")

      expect(button).toHaveAttribute("aria-disabled", "true")
    })
  })
})
