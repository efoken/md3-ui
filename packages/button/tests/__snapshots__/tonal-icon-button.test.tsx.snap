// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TonalIconButton /> prop: selected should select the component 1`] = `
<button
  class="css-view-175oi2r r-touchAction-1otgn73 Pressable-root ButtonBase-root r-cursor-1loqt21 r-outlineWidth-10paoce r-transition-71lksq r-userSelect-lrvibr TonalIconButton-root r-alignItems-1awozwy r-backgroundColor-oc2s68 r-borderRadius-cpet4d r-height-eu3ka r-justifyContent-1777fci r-width-1aockid"
  role="button"
  tabindex="0"
  type="button"
>
  <div
    style="color: rgb(29, 25, 43); display: contents;"
  >
    <svg
      class="Icon-root r-flexShrink-1q142lx r-fontSize-1x35g6 r-height-1472mwg r-width-lrsllp"
      color="#1d192b"
      fill="currentColor"
      size="24"
      style="color: rgb(29, 25, 43);"
    />
  </div>
  <span
    class="ButtonBase-hover r-borderRadius-sqpuna r-bottom-1p0dtai r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af r-transition-sp8bj3 r-zIndex-1wyyakw"
    style="background-color: rgb(29, 25, 43); opacity: 0;"
  />
  <span
    class="ButtonBase-rippleContainer r-borderRadius-sqpuna r-bottom-1p0dtai r-left-1d2f490 r-overflow-1udh08x r-position-u8s1d r-right-zchlnj r-top-ipm5af r-zIndex-1wyyakw"
  />
</button>
`;

exports[`<TonalIconButton /> prop: selected should unselect the component 1`] = `
<button
  class="css-view-175oi2r r-touchAction-1otgn73 Pressable-root ButtonBase-root r-cursor-1loqt21 r-outlineWidth-10paoce r-transition-71lksq r-userSelect-lrvibr TonalIconButton-root r-alignItems-1awozwy r-backgroundColor-cvqy8h r-borderRadius-cpet4d r-height-eu3ka r-justifyContent-1777fci r-width-1aockid"
  role="button"
  tabindex="0"
  type="button"
>
  <div
    style="color: rgb(73, 69, 79); display: contents;"
  >
    <svg
      class="Icon-root r-flexShrink-1q142lx r-fontSize-1x35g6 r-height-1472mwg r-width-lrsllp"
      color="#49454f"
      fill="currentColor"
      size="24"
      style="color: rgb(73, 69, 79);"
    />
  </div>
  <span
    class="ButtonBase-hover r-borderRadius-sqpuna r-bottom-1p0dtai r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af r-transition-sp8bj3 r-zIndex-1wyyakw"
    style="background-color: rgb(73, 69, 79); opacity: 0;"
  />
  <span
    class="ButtonBase-rippleContainer r-borderRadius-sqpuna r-bottom-1p0dtai r-left-1d2f490 r-overflow-1udh08x r-position-u8s1d r-right-zchlnj r-top-ipm5af r-zIndex-1wyyakw"
  />
</button>
`;

exports[`<TonalIconButton /> should render 1`] = `
<button
  class="css-view-175oi2r r-touchAction-1otgn73 Pressable-root ButtonBase-root r-cursor-1loqt21 r-outlineWidth-10paoce r-transition-71lksq r-userSelect-lrvibr TonalIconButton-root r-alignItems-1awozwy r-backgroundColor-oc2s68 r-borderRadius-cpet4d r-height-eu3ka r-justifyContent-1777fci r-width-1aockid"
  role="button"
  tabindex="0"
  type="button"
>
  <div
    style="color: rgb(29, 25, 43); display: contents;"
  >
    <svg
      class="Icon-root r-flexShrink-1q142lx r-fontSize-1x35g6 r-height-1472mwg r-width-lrsllp"
      color="#1d192b"
      data-testid="icon"
      fill="currentColor"
      size="24"
      style="color: rgb(29, 25, 43);"
    />
  </div>
  <span
    class="ButtonBase-hover r-borderRadius-sqpuna r-bottom-1p0dtai r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af r-transition-sp8bj3 r-zIndex-1wyyakw"
    style="background-color: rgb(29, 25, 43); opacity: 0;"
  />
  <span
    class="ButtonBase-rippleContainer r-borderRadius-sqpuna r-bottom-1p0dtai r-left-1d2f490 r-overflow-1udh08x r-position-u8s1d r-right-zchlnj r-top-ipm5af r-zIndex-1wyyakw"
  />
</button>
`;
